core/App/ehsm_napi.cpp

```diff
@@ -375,26 +374,97 @@ char* NAPI_GenerateDataKey(const char* cmk_base64,
     plaintext_base64 = base64_encode(plaint_datakey.data, plaint_datakey.datalen);
     ciphertext_base64 = base64_encode(cipher_datakey.data, cipher_datakey.datalen);
     if(plaintext_base64.size() > 0 ){
-        retJsonObj.addData("plaintext_base64", plaintext_base64);
-
-        if(ciphertext_base64.size() > 0){
-            retJsonObj.addData("ciphertext_base64", ciphertext_base64);
+        retJsonObj.addData("plaintext_base64", plaintext_base64); 
+		if(ciphertext_base64.size() > 0){
+			retJsonObj.addData("ciphertext_base64", ciphertext_base64);
+			SAFE_FREE(masterkey.keyblob);
+			SAFE_FREE(plaint_datakey.data);
+			SAFE_FREE(cipher_datakey.data);
+			return retJsonObj.toChar();
+		}
+    }
+    
```

Comment:
this code can be refined as this below:
"
if((plaintext_base64.size() > 0)  && (ciphertext_base64.size() > 0) ){
        retJsonObj.addData("plaintext_base64", plaintext_base64); 
        retJsonObj.addData("ciphertext_base64", ciphertext_base64);
    }
"

and same with the other functions,  please remove the duplicate codes when success return.
"
SAFE_FREE(xxx);
SAFE_FREE(xxx);
return retJsonObj.toChar();